File: app/src/main/java/com/ShaHar91/poems/ui/home/poem/adapter/PoemDetailAdapterController.java
package com.shahar91.poems.ui.home.poem.adapter;

import android.content.Context;

import com.airbnb.epoxy.AutoModel;
import com.airbnb.epoxy.Typed2EpoxyController;
import com.airbnb.epoxy.Typed3EpoxyController;
import com.shahar91.poems.data.models.Poem;
import com.shahar91.poems.data.models.Review;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public class PoemDetailAdapterController extends Typed2EpoxyController<Poem, Review> implements PoemNoReviewModel.Listener, PoemReviewModel.Listener {
    public interface Listener {
        void onRatingBarTouched(float rating);

        void onEditReviewClicked(Review review);
    }

    @AutoModel
    PoemDetailDataModel_ poemDetail;
    @AutoModel
    PoemNoReviewModel_ poemNoReviewModel;
    @AutoModel
    PoemReviewModel_ poemReviewModel;
    @AutoModel
    PoemGlobalRatingModel_ poemGlobalRatingModel;

    private final Context context;
    private final Listener listener;

    public PoemDetailAdapterController(Context context, Listener listener) {
        this.context = context;
        this.listener = listener;
    }

    @Override
    protected void buildModels(@Nonnull Poem poem, @Nullable Review ownReview) {
        poemDetail.poem(poem)
                .addTo(this);

        poemNoReviewModel
                .listener(this)
                .addIf(ownReview == null, this);

        if (ownReview != null) {
            poemReviewModel
                    .review(ownReview)
                    .listener(this)
                    .addTo(this);
        }

        if (!poem.getShortReviewList().isEmpty()) {
            poemGlobalRatingModel
                    .averageRating(poem.getAverageRating())
                    .totalRatingCount(poem.getTotalRatingCount())
                    .addTo(this);

            // TODO: only retrieve the 5 top reviews... create new screen to see a paged list of all reviews!!!
            for (Review review: poem.getShortReviewList()) {
                new PoemReviewModel_()
                        .id(review.getId())
                        .review(review)
                        .listener(this)
                        .addTo(this);
            }
        }
    }

    @Override
    public void onRatingBarTouched(float rating) {
        listener.onRatingBarTouched(rating);
    }

    @Override
    public void onEditReviewClicked(@NotNull Review review) {
        listener.onEditReviewClicked(review);
    }
}

File: app/src/main/java/com/ShaHar91/poems/ui/home/poem/adapter/PoemDetailAdapterController.kt
package com.shahar91.poems.ui.home.poem.adapter

import android.content.Context
import com.airbnb.epoxy.Typed2EpoxyController
import com.shahar91.poems.data.models.Poem
import com.shahar91.poems.data.models.Review

class PoemDetailAdapterController(private val context: Context,
    private val listener: Listener) :
    Typed2EpoxyController<Poem, Review?>() {
    interface Listener {
        fun onRatingBarTouched(rating: Float)
        fun onEditReviewClicked(review: Review)
        fun onDeleteReviewClicked(reviewId: String)
    }

    private val reviewListener = object : PoemReviewModel.Listener {
        override fun onEditReviewClicked(review: Review) {
            listener.onEditReviewClicked(review)
        }

        override fun onDeleteReviewClicked(reviewId: String) {
            listener.onDeleteReviewClicked(reviewId)
        }
    }

    override fun buildModels(poem: Poem, ownReview: Review?) {
        poemDetailData {
            this.id(poem._id)
            this.poem(poem)
        }

        if (ownReview == null) {
            poemNoReview {
                this.id("OwnReview")
                this.listener(object : PoemNoReviewModel.Listener {
                    override fun onRatingBarTouched(rating: Float) {
                        listener.onRatingBarTouched(rating)
                    }
                })
            }
        } else {
            poemReview {
                this.id(ownReview._id)
                this.review(ownReview)
                this.listener(reviewListener)
            }
        }

        if (poem.shortReviewList.isNotEmpty()) {
            //TODO: issues exist when adding or removing own review not updating the global review layout due to epoxy's hatefull state and immutable shit....
            poemGlobalRating {
                this.id(poem._id)
                this.averageRating(poem.averageRating)
                this.totalRatingCount(poem.totalRatingCount)
            }

            // TODO: only retrieve the 5 top reviews...(DONE) create new screen to see a paged list of all reviews!!! (TODO)
            poem.shortReviewList.forEach {
                poemReview {
                    this.id(it._id)
                    this.review(it)
                    this.listener(reviewListener)
                }
            }
        }
    }
}